File: evidence/v1_1/ai_eval/subjects/pilot/r1/block_scan/cpp/main.cpp
#include <algorithm>
#include <barrier>
#include <cstdint>
#include <iostream>
#include <thread>
#include <vector>

namespace {

constexpr std::size_t kBlockSize = 256;

// Number of Hillis-Steele scan steps needed so that 2^steps >= len.
std::size_t stepsFor(std::size_t len) {
  std::size_t steps = 0;
  while ((std::size_t{1} << steps) < len) ++steps;
  return steps;
}

}  // namespace

int main() {
  std::ios::sync_with_stdio(false);
  std::cin.tie(nullptr);

  std::size_t n;
  if (!(std::cin >> n)) return 0;

  std::vector<std::uint64_t> input(n);
  for (std::size_t i = 0; i < n; ++i) {
    unsigned long long v;
    std::cin >> v;
    input[i] = static_cast<std::uint64_t>(v);
  }

  if (n == 0) return 0;

  std::vector<std::uint64_t> output(n);
  const std::size_t numBlocks = (n + kBlockSize - 1) / kBlockSize;

  // The team: each thread scans a contiguous lane of a block, then the
  // team combines the per-lane totals across the shared array with a
  // barrier-separated step scan, exactly as a GPU block would combine
  // the partial sums produced by its lanes. A small, fixed team size
  // avoids pathological OS scheduling when it happens to match the
  // machine's core count exactly.
  const std::size_t numThreads = std::min<std::size_t>({8, kBlockSize, n});

  std::barrier sync_point(static_cast<std::ptrdiff_t>(numThreads));

  // Shared scratch array of per-lane block totals (the array the team
  // cooperatively scans in steps).
  std::uint64_t laneA[kBlockSize];
  std::uint64_t laneB[kBlockSize];
  std::uint64_t runningOffset = 0;

  auto worker = [&](std::size_t t) {
    const std::size_t chunk = (kBlockSize + numThreads - 1) / numThreads;

    for (std::size_t b = 0; b < numBlocks; ++b) {
      const std::size_t base = b * kBlockSize;
      const std::size_t blockLen = std::min(kBlockSize, n - base);

      const std::size_t start = std::min(t * chunk, blockLen);
      const std::size_t end = std::min(start + chunk, blockLen);

      std::uint64_t lane_sum = 0;
      for (std::size_t i = start; i < end; ++i) {
        lane_sum += input[base + i];
        output[base + i] = lane_sum;  // local (within-lane) inclusive scan
      }
      laneA[t] = lane_sum;

      sync_point.arrive_and_wait();

      // Scan the per-lane totals held in the shared array, in steps,
      // with every thread of the team waiting at a barrier before any
      // thread starts the next step.
      std::uint64_t *cur = laneA;
      std::uint64_t *nxt = laneB;
      const std::size_t steps = stepsFor(numThreads);
      for (std::size_t s = 0; s < steps; ++s) {
        const std::size_t d = std::size_t{1} << s;
        std::uint64_t val = cur[t];
        if (t >= d) val += cur[t - d];
        nxt[t] = val;
        sync_point.arrive_and_wait();
        std::swap(cur, nxt);
      }

      const std::uint64_t offset = runningOffset + (cur[t] - lane_sum);
      for (std::size_t i = start; i < end; ++i) output[base + i] += offset;

      sync_point.arrive_and_wait();

      if (t == 0) runningOffset += cur[numThreads - 1];
    }
  };

  std::vector<std::thread> threads;
  threads.reserve(numThreads);
  for (std::size_t t = 0; t < numThreads; ++t) threads.emplace_back(worker, t);
  for (auto &th : threads) th.join();

  std::string out;
  out.reserve(n * 12);
  for (std::size_t i = 0; i < n; ++i) {
    out += std::to_string(output[i]);
    out += '\n';
  }
  std::cout << out;

  return 0;
}

File: evidence/v1_1/ai_eval/subjects/pilot/r1/block_scan/rust/main.rs
use std::io::{self, Read};
use std::sync::atomic::{AtomicU64, Ordering};
use std::sync::Barrier;
use std::thread;

const BLOCK_SIZE: usize = 256;

// Reads the whole of standard input.
fn read_input() -> Vec<u8> {
    let mut bytes = Vec::new();
    io::stdin().read_to_end(&mut bytes).unwrap();
    bytes
}

// The doubling offsets used by a Hillis-Steele inclusive scan of a block of
// length m: 1, 2, 4, ... up to (but excluding) m. Depends only on m, so every
// lane in the team computes the same sequence independently.
fn scan_offsets(m: usize) -> Vec<usize> {
    let mut offsets = Vec::new();
    let mut off = 1usize;
    while off < m {
        offsets.push(off);
        off *= 2;
    }
    offsets
}

fn main() {
    let input = read_input();
    let text = String::from_utf8(input).unwrap();
    let mut nums = text.split_ascii_whitespace().map(|s| s.parse::<u64>().unwrap());

    let n = nums.next().unwrap_or(0) as usize;
    let values: Vec<u64> = (0..n).map(|_| nums.next().unwrap()).collect();

    if n == 0 {
        return;
    }

    let num_blocks = (n + BLOCK_SIZE - 1) / BLOCK_SIZE;
    // Team size: enough lanes to cover a block's elements, but capped to the
    // machine's parallelism so the barrier isn't shared by far more threads
    // than can actually run at once. Lanes are reused for every block
    // instead of respawning threads each time; a lane covers several
    // elements per step when the team is smaller than the block.
    let cpus = thread::available_parallelism().map(|p| p.get()).unwrap_or(1);
    let team_size = std::cmp::min(BLOCK_SIZE, n).min(cpus.max(1));

    let results: Vec<AtomicU64> = (0..n).map(|_| AtomicU64::new(0)).collect();
    // Two scratch buffers the whole team shares, ping-ponged between scan
    // steps so a step's reads never race with that same step's writes.
    let buf_a: Vec<AtomicU64> = (0..BLOCK_SIZE).map(|_| AtomicU64::new(0)).collect();
    let buf_b: Vec<AtomicU64> = (0..BLOCK_SIZE).map(|_| AtomicU64::new(0)).collect();
    let barrier = Barrier::new(team_size);
    let running_offset = AtomicU64::new(0);

    let values_ref = &values;
    let results_ref = &results;
    let buf_a_ref = &buf_a;
    let buf_b_ref = &buf_b;
    let barrier_ref = &barrier;
    let running_offset_ref = &running_offset;

    thread::scope(|s| {
        for lane in 0..team_size {
            s.spawn(move || {
                for block_idx in 0..num_blocks {
                    let start = block_idx * BLOCK_SIZE;
                    let end = std::cmp::min(start + BLOCK_SIZE, n);
                    let m = end - start;

                    // Load this block's values into the shared scratch array.
                    let mut idx = lane;
                    while idx < m {
                        buf_a_ref[idx].store(values_ref[start + idx], Ordering::SeqCst);
                        idx += team_size;
                    }
                    barrier_ref.wait();

                    let offsets = scan_offsets(m);
                    for (step, &off) in offsets.iter().enumerate() {
                        let (read_buf, write_buf) = if step % 2 == 0 {
                            (buf_a_ref, buf_b_ref)
                        } else {
                            (buf_b_ref, buf_a_ref)
                        };
                        let mut idx = lane;
                        while idx < m {
                            let own = read_buf[idx].load(Ordering::SeqCst);
                            let new_val = if idx >= off {
                                own + read_buf[idx - off].load(Ordering::SeqCst)
                            } else {
                                own
                            };
                            write_buf[idx].store(new_val, Ordering::SeqCst);
                            idx += team_size;
                        }
                        barrier_ref.wait();
                    }

                    let result_buf = if offsets.len() % 2 == 0 {
                        buf_a_ref
                    } else {
                        buf_b_ref
                    };

                    // Offset each element by the total of all earlier blocks.
                    let offset_val = running_offset_ref.load(Ordering::SeqCst);
                    let mut idx = lane;
                    while idx < m {
                        let scanned = result_buf[idx].load(Ordering::SeqCst);
                        results_ref[start + idx].store(scanned + offset_val, Ordering::SeqCst);
                        idx += team_size;
                    }
                    barrier_ref.wait();

                    // One lane folds this block's total into the running
                    // offset for the next block, only after every lane has
                    // already read (and used) the old value above.
                    if lane == 0 {
                        let block_total = result_buf[m - 1].load(Ordering::SeqCst);
                        running_offset_ref.fetch_add(block_total, Ordering::SeqCst);
                    }
                    barrier_ref.wait();
                }
            });
        }
    });

    let mut out = String::with_capacity(n * 12);
    for r in &results {
        out.push_str(&r.load(Ordering::SeqCst).to_string());
        out.push('\n');
    }
    print!("{}", out);
}
